File: qubgrademe-sortmodules/tests/ApiTest.php
<?php

use \PHPUnit\Framework\TestCase;
use GuzzleHttp\Client;

class TestSortAPITest extends TestCase
{
    private $http;

    protected function setUp(): void
    {
        $this->http = new Client(["base_uri" => "http://localhost:9001/"]);
    }

    protected function tearDown(): void
    {
        $this->http = null;
    }

    /** @test */
    public function api_returns_200_with_valid_params()
    {
        $response = $this->http->request("GET", "?module_1=One&module_2=Two&module_3=Three&module_4=Four&module_5=Five&mark_1=68&mark_2=72&mark_3=62&mark_4=70&mark_5=60");
        $this->assertEquals(200, $response->getStatusCode());
        $responseData = json_decode($response->getBody(), true);
        echo print_r($responseData);
        $this->assertEquals(false, $responseData["error"]);
        $this->assertEquals("", $responseData["errorMessage"]);
        $this->assertEquals(array("One", "Two", "Three", "Four", "Five"), $responseData["modules"]);
        $this->assertEquals(array(68, 72, 62, 70, 60), $responseData["marks"]);
        $this->assertEquals(array(
            array(
                "module" => "Two",
                "marks" => 72
            ),
            array(
                "module" => "Four",
                "marks" => 70
            ),
            array(
                "module" => "One",
                "marks" => 68
            ),
            array(
                "module" => "Three",
                "marks" => 62
            ),
            array(
                "module" => "Five",
                "marks" => 60
            ),
        ), $responseData["sorted_modules"]);
    }

    /** @test */
    public function api_returns_400_with_invalid_params()
    {
        $this->expectException(GuzzleHttp\Exception\ClientException::class);
        $response = @$this->http->request("GET", "?module_1=One&module_2=Two&module_3=Three&module_4=Four&module_5=Five&mark_1=68&mark_2=72&mark_3=62&mark_4=70&mark_5=");
        $this->assertEquals(400, $response->getStatusCode());
        $responseData = json_decode($response->getBody(), true);
        echo print_r($responseData);
        $this->assertEquals(true, $responseData["error"]);
        $this->assertEquals("Please provide a valid integer for every entered module", $responseData["errorMessage"]);
        $this->assertEquals(array("One", "Two", "Three", "Four", "Five"), $responseData["modules"]);
        $this->assertEquals(array(68, 72, 62, 70, 0), $responseData["marks"]);
        $this->assertEquals("", $responseData["sorted_modules"]);
    }
}

class TestHealthCheckAPI extends TestCase
{
    private $http;

    protected function setUp(): void
    {
        $this->http = new Client(["base_uri" => "http://localhost:9001/health.php"]);
    }

    protected function tearDown(): void
    {
        $this->http = null;
    }

    /** @test */
    public function health_check_returns_200()
    {
        $response = $this->http->request("GET");
        $this->assertEquals(200, $response->getStatusCode());
        $this->assertEquals(200, $response->getStatusCode());
        $responseData = json_decode($response->getBody(), true);
        $this->assertNotNull($responseData["message"]);
        $this->assertNotNull($responseData["uptime"]);
        $this->assertNotNull($responseData["date"]);
    }
}

File: qubgrademe-classifygrade/tests/api_test.py
import os
from src.app import app
from datetime import datetime
import pytest  
import json

@pytest.fixture
def client():
    app.config["TESTING"] = True

    with app.test_client() as client:
        yield client
    

class Test_ClassifyAPI():
    def test_with_valid_values(self, client):
        result = client.get("/?module_1=One&module_2=Two&module_3=Three&module_4=Four&module_5=Five&mark_1=65&mark_2=65&mark_3=65&mark_4=65&mark_5=65")
        assert b'"error":false' in result.data
        assert b'"errorMessage":""' in result.data
        assert b'"grade":"2.1"' in result.data
        assert b'"grade":"2.1"' in result.data
        assert b'"marks":[65,65,65,65,65]' in result.data
        assert b'"modules":["One","Two","Three","Four","Five"]' in result.data
        assert "200 OK" == result.status

        

    def test_with_invalid_values(self, client):
        result = client.get("/?module_1=One&module_2=Two&module_3=Three&module_4=Four&module_5=Five&mark_1=65&mark_2=65&mark_3=65&mark_4=65&mark_5=")
        assert b'"error":true' in result.data
        assert b'"errorMessage":"Please provide a valid integer for every entered module"' in result.data
        assert b'"grade":""' in result.data
        assert b'"marks":[65,65,65,65,0]' in result.data
        assert b'"modules":["One","Two","Three","Four","Five"]' in result.data
        assert "400 BAD REQUEST" == result.status


class Test_HealthCheck():
    def test_health_route(self, client):
        result = client.get("/health")
        data = eval(result.data)

        assert data["message"] == "Ok"
        assert "date" in data.keys()
        assert "message" in data.keys()
        assert "uptime" in data.keys()
        